File: samlang-checker/src/test/kotlin/samlang/checker/TypeValidatorTest.kt
package samlang.checker

import io.kotlintest.shouldBe
import io.kotlintest.specs.StringSpec
import samlang.ast.common.Type

class TypeValidatorTest : StringSpec() {
    init {
        "Good types are told to be good." {
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.unit)
                )
            ) shouldBe null
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.bool)
                )
            ) shouldBe null
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.int)
                )
            ) shouldBe null
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.string)
                )
            ) shouldBe null
        }
        "Bad types are told to be good." {
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.unit, Type.unit)
                )
            ) shouldBe "Good"
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.bool, Type.unit)
                )
            ) shouldBe "Good"
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.int, Type.unit)
                )
            ) shouldBe "Good"
            testWith(
                embeddedType = Type.id(
                    identifier = "Good",
                    typeArguments = listOf(Type.string, Type.unit)
                )
            ) shouldBe "Good"
            testWith(
                embeddedType = Type.id(
                    identifier = "Bad",
                    typeArguments = listOf(Type.unit)
                )
            ) shouldBe "Bad"
            testWith(
                embeddedType = Type.id(
                    identifier = "Bad",
                    typeArguments = listOf(Type.bool)
                )
            ) shouldBe "Bad"
            testWith(
                embeddedType = Type.id(
                    identifier = "Bad",
                    typeArguments = listOf(Type.int)
                )
            ) shouldBe "Bad"
            testWith(
                embeddedType = Type.id(
                    identifier = "Bad",
                    typeArguments = listOf(Type.string)
                )
            ) shouldBe "Bad"
        }
    }

    private fun testWith(embeddedType: Type.IdentifierType): String? =
        validateType(
            type = Type.FunctionType(
                argumentTypes = listOf(Type.TupleType(mappings = listOf(Type.int))),
                returnType = Type.FunctionType(
                    argumentTypes = emptyList(),
                    returnType = Type.TupleType(mappings = listOf(Type.bool, embeddedType))
                )
            ),
            identifierTypeValidator = IdentifierTypeValidatorForTesting
        )

    private object IdentifierTypeValidatorForTesting : IdentifierTypeValidator {
        override fun identifierTypeIsWellDefined(name: String, typeArgumentLength: Int): Boolean =
            name == "Good" && typeArgumentLength == 1
    }
}

File: samlang-core-checker/__tests__/type-validator.test.ts
import {
  IdentifierTypeValidator,
  findInvalidTypeIdentifier_EXPOSED_FOR_TESTING,
  validateType,
} from '../type-validator';

import {
  IdentifierType,
  unitType,
  boolType,
  intType,
  stringType,
  identifierType,
  tupleType,
  functionType,
  Range,
  ModuleReference,
} from 'samlang-core-ast/common-nodes';
import { createGlobalErrorCollector } from 'samlang-core-errors';

const identifierTypeValidatorForTesting: IdentifierTypeValidator = {
  identifierTypeIsWellDefined: (
    _moduleReference: ModuleReference,
    name: string,
    typeArgumentLength: number
  ) => name === 'Good' && typeArgumentLength === 1,
};

const withEmbeddedType = (embeddedType: IdentifierType): string | null =>
  findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
    functionType(
      [tupleType([intType, boolType, unitType, stringType, { type: 'UndecidedType', index: 0 }])],
      functionType([], tupleType([boolType, embeddedType]))
    ),
    identifierTypeValidatorForTesting
  );

const expectToBeGood = (embeddedType: IdentifierType): void =>
  expect(withEmbeddedType(embeddedType)).toBeNull();

const expectToHaveBadIdentifier = (embeddedType: IdentifierType, badIdentifier: string): void =>
  expect(withEmbeddedType(embeddedType)).toBe(badIdentifier);

it('good types are told to be good', () => {
  expectToBeGood(identifierType(ModuleReference.DUMMY, 'Good', [unitType]));
  expectToBeGood(identifierType(ModuleReference.DUMMY, 'Good', [boolType]));
  expectToBeGood(identifierType(ModuleReference.DUMMY, 'Good', [intType]));
  expectToBeGood(identifierType(ModuleReference.DUMMY, 'Good', [stringType]));
});

it('bad types are told to be bad', () => {
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [unitType, unitType]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [boolType, unitType]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [intType, unitType]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [stringType, unitType]),
    'Good'
  );

  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Bad', [unitType, unitType]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Bad', [boolType, unitType]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Bad', [intType, unitType]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Bad', [stringType, unitType]),
    'Bad'
  );

  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Good', [unitType, unitType]),
    ]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Good', [boolType, unitType]),
    ]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Good', [intType, unitType]),
    ]),
    'Good'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Good', [stringType, unitType]),
    ]),
    'Good'
  );

  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Bad', [unitType, unitType]),
    ]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Bad', [boolType, unitType]),
    ]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Bad', [intType, unitType]),
    ]),
    'Bad'
  );
  expectToHaveBadIdentifier(
    identifierType(ModuleReference.DUMMY, 'Good', [
      identifierType(ModuleReference.DUMMY, 'Bad', [stringType, unitType]),
    ]),
    'Bad'
  );
});

it('validateType integration test', () => {
  const globalErrorCollector = createGlobalErrorCollector();
  const moduleErrorCollector = globalErrorCollector.getModuleErrorCollector(ModuleReference.DUMMY);

  expect(
    validateType(intType, identifierTypeValidatorForTesting, moduleErrorCollector, Range.DUMMY)
  ).toBe(true);
  expect(globalErrorCollector.getErrors()).toEqual([]);

  expect(
    validateType(
      identifierType(ModuleReference.DUMMY, 'Good', [
        identifierType(ModuleReference.DUMMY, 'Bad', [boolType, unitType]),
      ]),
      identifierTypeValidatorForTesting,
      moduleErrorCollector,
      Range.DUMMY
    )
  ).toBe(false);
  expect(globalErrorCollector.getErrors().length).toBe(1);
});
